### PR TITLE
improve: clean up some string caches on stack in ptracer

### PR DESCRIPTION
**Pull request by Microwwbin**
## Changes

Try to hide some traces of zygisk.

## Why 

This looks like a "found injection" check point. You can simply validate it by return directly in entry.cpp#entry.

### CODE REVIEW

**Comment by ThePedroo** [comment]
Will look into it, thanks!

File: loader/src/ptracer/ptracer.cpp [hunk 1]
```diff
@@ -16,6 +16,15 @@
 
 #include "utils.hpp"
 
+static void clearStackStringData(int pid, struct user_regs_struct& regs, const char *path) {
+    auto len = strlen(path) + 1;
```
Comment (ThePedroo) [inline loader/src/ptracer/ptracer.cpp:20]: The goal of the project is to re-write it all to C99, so please avoid the use of C++ specific keywords.

File: loader/src/ptracer/ptracer.cpp [hunk 2]
```diff
@@ -16,6 +16,15 @@
...
+    char empty[len];
+    memset(empty, 0x1, len);
+    empty[len] = '\0';
+    regs.REG_SP += len;
```
Comment (ThePedroo) [inline loader/src/ptracer/ptracer.cpp:24]: Please break the line after and before regs.RED_SP

File: loader/src/ptracer/ptracer.cpp [hunk 3]
```diff
@@ -16,6 +16,15 @@
 
 #include "utils.hpp"
 
+static void clearStackStringData(int pid, struct user_regs_struct& regs, const char *path) {
```
Comment (ThePedroo) [inline loader/src/ptracer/ptracer.cpp:19]: `clearStackStringData` -> `hide_static_string` or `clear_static_string`

File: loader/src/ptracer/ptracer.cpp [hunk 4]
```diff
@@ -199,11 +209,13 @@ bool inject_on_main(int pid, const char *lib_path) {
...
 
     args.clear();
-    str = push_string(pid, regs, "entry");
+    const char *entry = "entry";
+    str = push_string(pid, regs, entry);
     args.push_back(remote_handle);
     args.push_back((long) str);
 
     uintptr_t injector_entry = remote_call(pid, regs, (uintptr_t)dlsym_addr, (uintptr_t)libc_return_addr, args);
+    clearStackStringData(pid, regs, entry);
```
Comment (ThePedroo) [inline loader/src/ptracer/ptracer.cpp:218]: Ditto

File: loader/src/ptracer/ptracer.cpp [hunk 5]
```diff
@@ -148,6 +157,7 @@ bool inject_on_main(int pid, const char *lib_path) {
     args.push_back((long) RTLD_NOW);
 
     uintptr_t remote_handle = remote_call(pid, regs, (uintptr_t)dlopen_addr, (uintptr_t)libc_return_addr, args);
+    clearStackStringData(pid, regs, lib_path);
```
Comment (ThePedroo) [inline loader/src/ptracer/ptracer.cpp:160]: Break the line after the func call

**Review by ThePedroo — CHANGES_REQUESTED** [review]
Quick look, only syntax changes. However I'd like, if possible, a simple POC of how this (stack /static strings) could be detected

**Comment by Microwwbin** [comment]
**Test on A13**
Dump the stack data with follow code:
```cpp
    std::ofstream dump_data("/data/data/<package_name>/cache/dump", std::ios::binary);
    if (dump_data.is_open())
        for (auto &map: lsplt::MapInfo::Scan()) {
            if ((/*map.path.find("anon:stack_and_tls") != std::string::npos ||*/
                        map.path.find("[stack]") != std::string::npos) &&
                (map.perms & PROT_READ)) {
                auto address = reinterpret_cast<char *>(map.start);
                for (; address < (void *) map.end; address++) {
                    if (*address) dump_data.write(reinterpret_cast<char *>(address), 1);
                }
                LOGD("has dumped %s", map.path.c_str());
            }
        }
    else
        LOGE("not open");
    dump_data.flush();
    dump_data.close();
```
In the dump file, would see:
image

**Comment by Microwwbin** [comment]
**Test on A13**
Dump the stack data with follow code:
```cpp
    std::ofstream dump_data("/data/data/<package_name>/cache/dump", std::ios::binary);
    if (dump_data.is_open())
        for (auto &map: lsplt::MapInfo::Scan()) {
            if ((/*map.path.find("anon:stack_and_tls") != std::string::npos ||*/
                        map.path.find("[stack]") != std::string::npos) &&
                (map.perms & PROT_READ)) {
                auto address = reinterpret_cast<char *>(map.start);
                for (; address < (void *) map.end; address++) {
                    if (*address) dump_data.write(address, 1);
                }
                LOGD("has dumped %s", map.path.c_str());
            }
        }
    else
        LOGE("not open");
    dump_data.flush();
    dump_data.close();
```
In the dump file, would see:
60cd0d65-f1a4-48d3-814a-35da547f42a9

**Comment by Microwwbin** [comment]
**Test on A13, confirmed this is one of "found injection" check point​**
Dump the stack data with follow code:​
```cpp
    std::ofstream dump_data("/data/data/<package_name>/cache/dump", std::ios::binary);
    if (dump_data.is_open())
        for (auto &map: lsplt::MapInfo::Scan()) {
            if ((/*map.path.find("anon:stack_and_tls") != std::string::npos ||*/
                        map.path.find("[stack]") != std::string::npos) &&
                (map.perms & PROT_READ)) {
                auto address = reinterpret_cast<char *>(map.start);
                for (; address < (void *) map.end; address++) {
                    if (*address) dump_data.write(address, 1);
                }
                LOGD("has dumped %s", map.path.c_str());
            }
        }
    else
        LOGE("not open");
    dump_data.flush();
    dump_data.close();
```
In the dump file, would see:
image

**Comment by ThePedroo** [comment]
If we mmap to then munmap, theorically yes. This could possibly be better as it removes entirely from memory maps, supposedly. Any thoughts on this, @Microwwbin?

**Comment by ThePedroo** [comment]
Also, I don`t think debug_ramdisk string is coming from ReZygisk -- it has been a lot of time since we don`t use it anymore

**Comment by ThePedroo** [comment]
??? I did not even touch this PR.

**Comment by JingMatrix** [comment]
I didn't find the string `debug_r` in the stack.

**Pull request closed without merging**